File: Algorithm/leetcode_155.go
/**
 * File              : leetcode_155.go
 * Author            : Andy
 * Date              : 27.07.2022
 * Last Modified Date: 27.07.2022
 * Last Modified By  : Andy
 */


/**
主体思路：保存两个栈，一个正常存数据，另一个存最小前缀栈，即当前值是前面几个数中的最小值
getMin即是从这个最小前缀中取值即可
*/

type MinStack struct {
    stack []int 
    minStack []int
}

func Constructor() MinStack{
    return MinStack{
        stack: []int{},
        minStack: []int{math.MaxInt64},
    } 
}

func (this *MinStack) Push(val int){
    this.stack = append(this.stack, val)
    top := this.minStack[len(this.minStack)-1]
    this.minStack = append(this.minStack, min(val, top))
}

func (this *MinStack) Pop() {
    this.stack = this.stack[:len(this.stack)-1]
    this.minStack = this.minStack[:len(this.minStack)-1]
}

func (this *MinStack) top() {
    return this.stack[len(this.stack)-1]
}

func (this *MinStack) getTop(){
    return this.minStack[len(this.minStack)-1]
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

File: Algorithm/leetcode_155.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_155.py
# Author            : Andy
# Date              : 27.07.2022
# Last Modified Date: 27.07.2022
# Last Modified By  : Andy

# 主体思路是用一个额外的栈存储一个最小前缀值，即当前数是前面n个数中最小值
# 在pop时两个栈同时pop
# 在push时，一个正常存储，另一个则只存当前值与栈底元素中最小的那个值

class Minstack:
    def __init__(self):
        self.__stack = []
        self.__min_stack = []

    def push(self, val:int) ->None:
        self.__stack.append(val)
        if not self.__min_stack:
            self.__min_stack.append(val)
        else:
            self.__min_stack.append(min(val, self.__min_stack[-1]))

    def pop(self) ->None:
        self.__stack.pop()
        self.__min_stack.pop()

    def top(self) ->int:
        return self.__min_stack[-1]
            
    def getMin(self) -> int:
        return self.__min_stack[-1]
